### ISSUE TITLE
Support for RegexSet-like cases

### ISSUE THREAD

**Issue by RReverser** [BurntSushi/regex-automata#4]
I wonder if it would be possible to add support for adding multiple regexes one-by-one to an existing DFA to support cases like `regex::RegexSet` where you can distinguish which ones actually matched.

In terms of a minimal addition to the API, I imagine this would look like disabling/delaying `byte_classes` optimisation for such case and providing a method like `add_pattern(re: &str) -> Result<S>` that would allow consumer to get `StateID` for each regex individually.

Beyond that, `DFA` trait already exposes required methods for manual state walking so consumer could walk through the input byte-by-byte and, upon reaching the match, compare the last returned state to known regexes, so no change here is necessary (although convenience methods might be added in the future).

What do you think?

**Comment by BurntSushi** [comment]
Yes, I do want to do this. It looks like I wrote this in the README:

> Add support for regex sets. It should be possible to do this by "simply" introducing more match states. I think we can also report the positions at each match, similar to how Aho-Corasick works. I think the long pole in the tent here is probably the API design work and arranging it so that we don't introduce extra overhead into the non-regex-set case without duplicating a lot of code. It seems doable.

In particular, one of my personal pet use cases for this is for implementing various Unicode algorithms, e.g., https://github.com/BurntSushi/bstr/blob/dc8f855e58754ca972bfefed27adf927c6aac134/scripts/regex/word.sh#L25-L31 (Where a Unicode codepoint can appear in one of several properties, where you want to know which property it's in.)

I do think this is probably a fair amount of work, although I can't remember the original implementation path I had envisioned for this. I think the _core_ change necessary is augmenting the `Match` state with an index corresponding to the pattern that it belongs to:

https://github.com/BurntSushi/regex-automata/blob/a04de1f456a319fd838693dfd03af709d66e995f/src/nfa.rs#L51-L53

At that point, there should be some small changes required to support it in determinization and potentially minimization. But I don't think it's anything crazy.

At that point, there need to be additional methods that expose a new `RegexSet`-like API. At the lowest level, we'll need some API that, given a state ID to a match state, returns the index of the corresponding pattern.

I think one of my chief concerns here was how to do all of this such that it doesn't result in doubling portions of the code. I don't know how closely you've look at the implementation, but there's quite a bit of "near" duplication depending on the structure of the automaton. And unfortunately, there are some optimizations specifically with respect to querying whether a state is a match or not, and my hope is that we can keep those with `RegexSet`.

`RegexSets` will likely also require some modification to the serialization routines, although I generally expect that to be straight-forward.

While this is something I want, I'm not sure when I'll get to it. You could try to take a crack at this if you need this soon, but it would probably be good to draw up a more concrete implementation path. I'd be happy to try to mentor this work synchronously via chat if that helps. If this isn't urgent, then you could also just wait for me to do this. :-)

**Comment by RReverser** [comment]
> Yes, I do want to do this. It looks like I wrote this in the README:

I did see that you want to do this in README, but thought would open an issue with some ideas about how first stage of this could be added before the separate "nice" RegexSet API.

> You could try to take a crack at this if you need this soon

"need" is a bit of a strong word at the moment - I don't need it urgently, but it definitely looks like a promising path to migrate some of the old C / Ragel code to Rust. I guess I can wait :)

**Comment by awygle** [comment]
I'm interested in working on this. I'm curious what your plan for implementation was - I've come up with a few approaches but they all have one or more sticking points.

It's extremely easy to add alternations to `Hir` before even compiling to `NFA`, which gives all the functionality of `RegexSet` except knowing which pattern matched. Unfortunately there's no obvious way to plumb that information in at this level.

At the `NFA` level, it seems possible to augment the `NFACompiler` to provide some kind of `compile_multiple` method. This would necessarily break the currently stated invariant of only one `Match` state per `NFA`, however. I don't know how big of a deal this would turn out to be...

At the `DFA` level, one could store a `Vec<StateID>` of pattern indexes for all match states, I suppose. Once you get down this far it's not clear to me how to tell the `Match` states of the different patterns apart.

Currently I'd think NFA would be the right level of abstraction for this, and I'll spend some time hacking in that direction, but some guidance would be appreciated 😅

**Comment by BurntSushi** [comment]
> It's extremely easy to add alternations to Hir before even compiling to NFA, which gives all the functionality of RegexSet except knowing which pattern matched. Unfortunately there's no obvious way to plumb that information in at this level.

Right, yes. You can't do this at the HIR level. I'm not sure I can think of any particular advantages of doing it at the HIR level anyway, although it is, I suppose, possible. But only after expanding the HIR itself. But when you [build a `RegexSet`](https://docs.rs/regex/1.2.1/regex/struct.RegexSetBuilder.html), you provide many different individual patterns, and they all each get their own HIR. So it's a bit of an impedance mismatch.

Also, to clarify, a `RegexSet` is _not_ just about know which pattern matched, but which pattern**s** matched. With an alternation, only one branch can match. But with a `RegexSet`, many branches can match simultaneously.

> At the NFA level, it seems possible to augment the NFACompiler to provide some kind of compile_multiple method. This would necessarily break the currently stated invariant of only one Match state per NFA, however. I don't know how big of a deal this would turn out to be...

Yes, indeed, this is the approach I had in mind. The `Match` instruction needs to grow an argument, probably corresponding to the index of the pattern that matched (starting from zero). I expect this will impact determinization in some subtle ways. You can see an example of that here: https://github.com/rust-lang/regex/blob/29f39b87213ccbd88c092dd48071ad60a96e7521/src/dfa.rs#L965-L969 --- There are other conditionals like that in that code.

To be honest, my suspicion is that making regex sets work at the automata level isn't going to be too difficult. The real challenge of this task is figuring out what the API looks like, and whether we can avoid duplicating too much code. I definitely encourage you to hack around and see what you can get working, but before spending too much time on the public API and other plumbing details, I'd also encourage you to come back here and propose an API. `regex-automata` doesn't have many dependents, so I am okay with doing breaking changes if that would make the API much better.

**Comment by BurntSushi** [comment]
Also, thanks very much for offering to work on this! :)

**Comment by awygle** [comment]
It does seem to be pretty easy to get working. I need to do some more testing but I think I already have it basically figured out.

As far as an API goes, I basically see two options. We can maintain a whole second parallel implementation of `DFA`, a `DFASet` or similar (and every concrete `DFA` type, I think), or we can always track the pattern indexes for all matches and make `DFA` methods like `find` return `Option<(usize, Vec<usize>)>`. One is going to be more efficient, the other is a lot less code duplication.

The "intermediate" types, like `dense::Builder` and `NFABuilder`, can unconditionally track pattern index without loss of runtime performance, and `RegexSet` is the primary API, so this is only really a question for the `DFA` types.

I haven't done a _ton_ of Rust API design, so maybe I'm missing something crucial, but those are the options as I see them.

**Comment by BurntSushi** [comment]
> It does seem to be pretty easy to get working. I need to do some more testing but I think I already have it basically figured out.

It might be wise to get this into a PR where it's at, just to make sure we're on the same page. Ideally, it would be mergeable as-is with no API changes or performance regressions. (I think.)

> As far as an API goes, I basically see two options. We can maintain a whole second parallel implementation of `DFA`, a `DFASet` or similar (and every concrete `DFA` type, I think), or we can always track the pattern indexes for all matches and make `DFA` methods like find return `Option<(usize, Vec<usize>)>`. One is going to be more efficient, the other is a lot less code duplication.

Okay, so let's try to untangle this a bit. The latter option, where we always return all match indices for all methods is a non-starter. Under _no circumstances_ can we regress performance of a single DFA in order to make multiple matching work. That's a hard constraint, even if it means lots of code duplication.

My _hope_ is that there's a third option. In particular, we leave the existing APIs alone for `DFA` but _add_ new methods that operate on multiple patterns, and thus may incur more overhead. My hope is that they could work by using some additional storage on the underlying dense/sparse automaton that maps match states to the corresponding pattern indices. This storage would simply not be used when using the single-pattern APIs.

 It'd probably be a good idea to write out the method signatures for these extra methods. Perhaps a source of inspiration could be the [`AhoCorasick::find_overlapping_iter`](https://docs.rs/aho-corasick/0.7.6/aho_corasick/struct.AhoCorasick.html#method.find_overlapping_iter) method, which is basically implementing what a regex set actually is: the ability to match multiple overlapping patterns. A `find_overlapping_iter` would probably be appropriate as an API item for a `RegexSet`, but we need lower level methods on the `DFA` trait to implement that. The aho-corasick crate implements that API, and the key thing that it tracks are state IDs. So that when you do overlapping searches, you carry over the state from your previous search.

Other APIs might be useful too, e.g., similar to the `regex::RegexSet` API.

In general, we should almost never be returning a `Vec`. If we need a `Vec` some where, the method should require it as a mutable argument and write to it.

> I haven't done a ton of Rust API design, so maybe I'm missing something crucial, but those are the options as I see them.

So basically, the primary reason why I haven't done this task yet is because the API design involved it is by far the hardest part, and I just don't have the bandwidth for it. I wish I could just cough up an API for you to implement and call it day, but I just don't have it in my head. :-(

**Comment by ids1024** [comment]
> A find_overlapping_iter would probably be appropriate as an API item for a RegexSet, but we need lower level methods on the DFA trait to implement that.

I'm not too familiar with this sort of thing, so I may be wrong. But all the DFA trait needs to provide is a way to find all the `StateID`s of the accept states that match, right? Like a `.find_overlapping_iter()` that returns `impl Iterator<Item=ID>`. The `RegexSet` would then contain the mapping from the state IDs to pattern indices.

Or I suppose if this is only useful to `RegexSet`, the needed functionality wouldn't even need to be a method of `DFA`; since like `.find()` it should be possible to implement just using DFA methods; so the code could go in `RegexSet`.

(And separately there would need need to be some mechanism to build a DFA from multiple patterns with a known relationship between the patterns and accept states; but that's not part of the DFA trait.)

**Comment by BurntSushi** [comment]
@ids1024 Yes, that may be sufficient. You can see how this works in aho-corasick with [`overlapping_find_at`](https://github.com/BurntSushi/aho-corasick/blob/4092c4b441200310a13c6edf59e19f30dc1e902b/src/automaton.rs#L484-L521) and [`get_match`](https://github.com/BurntSushi/aho-corasick/blob/4092c4b441200310a13c6edf59e19f30dc1e902b/src/automaton.rs#L92-L107).

**Comment by awygle** [comment]
I am still working on this although I know it must look like I've stopped :) life gets in the way as I'm sure you understand.

The `AhoCorasick` methods make a lot of sense to me in terms of what the `RegexSet` API wants to look like. I was never fully satisfied with `RegexSet` as implemented in `regex`, because it seems to me that (for example) the ability to find just the first match is often desirable.

My intent at this point is to make an implementation of `RegexSet` work which exposes methods equivalent in functionality (if not in name) to `is_match`, `find`, and `find_overlapping_iter` on the Aho-Corasick automatons found in `regex`, with reasonable semantics, no runtime performance regressions for single DFAs, and without worrying too much about code duplication. I'll then open a WIP pull request so we can discuss it, and once we're in agreement that the approach is reasonable, I'll try and collapse as much duplicated code as makes sense. Expect to see the PR sometime on or before Oct. 6, barring major disruptions. And of course, if you don't agree with this course of action, let me know - it's your crate after all.

**Comment by BurntSushi** [comment]
That sounds good. FWIW, just observing my own history, I've had a ton of problems getting large PRs merged. If you see any ooportunities for doing this piecemeal, that would probably improve the chances of success. Otherwise, your plan sounds good I think. Or at least, it roughly matches what is in my head.

**Comment by awygle** [comment]
The point's well-taken, but I'm not sufficiently versed in the subject matter to know what's worth merging before I have everything implemented I think - I'm sort of working things out as I go. For that reason, I'll look at splitting things up into smaller PRs from a "finished" branch rather than trying to submit piecemeal as I go.

**Comment by BurntSushi** [comment]
Yup that makes perfect sense, thanks. To be clear, I'm willing to entertain a giant PR, but just looking at my history, I still have giant PRs to review for other projects that I haven't gotten merged yet. (Some of them were submitted years ago.)

**Comment by BurntSushi** [comment]
As mentioned [here](https://github.com/BurntSushi/regex-automata/pull/5#issuecomment-712537096), I do have this working in my [branch containing an in-progress rewrite of most of the crate](https://github.com/BurntSushi/regex-automata/tree/ag/work). It will be a while yet before `0.2` release, but the functionality is there. And at least for DFAs, it is much more expressive than the regex crate's `RegexSet`. It supports reporting the offsets of matches, even when doing overlapping searches. And you can even provide a pattern ID to search to execute an anchored search of a specific pattern in the DFA.

**Comment by BurntSushi** [comment]
Note that this is done. `regex-automata 0.3` has first class support for multi-pattern regexes everywhere. That includes multiple regexes with capture groups.
